### PR TITLE
Add exchange ratio for ARS

### LINKED ISSUE #2283: Add a custom exchange rate for Argentine peso (ARS)
Hello! 

I have been working in the [Bitcoin Cash Argentina](https://bitcoincashargentina.com/) community to promote the adoption of Bitcoin Cash as a medium of exchange. We are focusing mainly in brick and mortar shops, but we have plans to reach more users. We promote the usage of various tools, including Electron Cash.

Due to problems in the exchange rates caused by local restrictions, we had to develop a custom system to enable users (both customers and merchants) to use the Argentine peso as a reference. The price displayed by most websites is incorrect and not the price a normal user would have to pay to acquire Bitcoin Cash.

I would like to ask you to implement this system in order to facilitate conversion for those merchants who do not have business smartphones and tablets (and thus are unable to use Register App) and would like to use computers. Our system is already usable in the Bitcoin.com wallet and in the Register App.

The rule is the following: We take the US dollar value in ARS from https://yadio.io/simple.html and then multiply it by the Bitcoin Cash value in dollars. Essentially:

	BCH price (in USD) × USD price (in ARS) = BCH price (in ARS)

I know that adding a custom exchange just for Argentina might be not standard. We try to deal with the restrictions the best that we can, and I hope this change might help in the work we are doing.

Thanks in advance!

### PR DESCRIPTION
**Pull request by schiab8**
Fixes #2283 
Adding custom "exchange" using bitstamp and yadio to get ARS rate

### CODE REVIEW

**Comment by mgiagante** [comment]
This change seems to solve the problem we have with the ARS/USD exchange rate, and as a consequence the one for ARS/BCH, since it's not the actual price people in Argentina have access to. Thanks for creating this PR @schiab8 !

**Comment by emergent-reasons** [comment]
I think you are missing at least this code inside `class BitstampYadio`:

```py
def history_ccys(self):
    return ['ARS']

```

If I do not add this code, the BitstampYadio does not appear as an exchange option.

**Comment by emergent-reasons** [comment]
Please also add all evidence to this MR that the rate provided by BitstampYadio is trusted. @schiab8 @mgiagante

**Comment by ianblas** [comment]
The government sets the official exchange rate artificially low.  Nobody uses that USD-ARS exchange rate. Everybody use the free market exchange rate.
Ian Blas:
The yadio rate is the correct exchange rate USD-ARS (https://yadio.io/simple.html)
bitcoin.com use yadio exchange rate USD-ARS.

Evidence:
Binance P2P USDT-ARS
Screenshot_20210815-031802_Binance

SatoshiTango (local crypto exchange) BCH-ARS 
Screenshot_20210815-031852_SatoshiTango

All the currently sources for BCH-ARS exchange rate are wrong. Because use the government official exchange rate USD-ARS

**Comment by schiab8** [comment]
> I think you are missing at least this code inside `class BitstampYadio`:
> 
> ```python
> def history_ccys(self):
>     return ['ARS']
> ```
> 
> If I do not add this code, the BitstampYadio does not appear as an exchange option.
> 
> Have you tested it? Does the option appear for you already?

Hi!
The option appears if you select ARS as currency first, then it will show BitstampYadio.
I took `Kraken` class as an example, which has defined `get_rates `only. Not sure what `history_ccys` is used for, i'm going to check that now.
image

**Comment by emergent-reasons** [comment]
I found the problem. There are two price modes:

a. current rate (this works with your patch)
b. historical rates: Turn this on with the option "Show history rates". This does not work with your patch for two reasons
    1. You need to add code for it
    2. I am not sure but I think bitstamp does not provide historical usd so it will not work for historical data. I think you can use coingecko instead for usd to get historical?

So I think b. needs to be investigated.

**Comment by schiab8** [comment]
Correct me if I'm wrong, but the only two exchanges that implement history_rates are Coincap and CoinGecko, I don't see the problem not having it

**Comment by cculianu** [comment]
Yeah I seem to recall that if the historical endpoint is missing, it still works, you just don't get historical data in the app, so ux is slightly degrated, but it still works.  I think it *should* be fine.

**Comment by schiab8** [comment]
Great! thank you

**Comment by emergent-reasons** [comment]
Does yadio provide historical data? If not, then ok. If they do, then this
pr should cover historical data too, no?

File: electroncash/exchange_rate.py [hunk 1]
```diff
@@ -261,6 +261,14 @@ def request_history(self, ccy):
         return dict([(dt.utcfromtimestamp(h[0]/1000).strftime('%Y-%m-%d'), h[1])
                      for h in history['prices']])
 
+        
+class BitstampYadio(ExchangeBase):
+    def get_rates(self, ccy):
+        json_usd = self.get_json('www.bitstamp.net', '/api/v2/ticker/bchusd')
+        json_ars = self.get_json('api.yadio.io', '/exrates/ARS')
+
```
Comment (cculianu) [inline electroncash/exchange_rate.py:269]:
```suggestion

```

File: electroncash/exchange_rate.py [hunk 2]
```diff
@@ -261,6 +261,14 @@ def request_history(self, ccy):
         return dict([(dt.utcfromtimestamp(h[0]/1000).strftime('%Y-%m-%d'), h[1])
                      for h in history['prices']])
 
+
```
Comment (cculianu) [inline electroncash/exchange_rate.py:264]:
```suggestion

```

**Review by cculianu — COMMENTED** [review]
Looks good. Tested. Yeah the official rates are too low. I guess this one has more "realistic" rates. 

Works as advertised. Note, that because the provider doesn't have historical data, the option only appears if you turn off history rates.

That's fine.  Lots of other rate services work that way.  

Left some whitespace-related suggested changes...

**Comment by cculianu** [comment]
> Does yadio provide historical data? If not, then ok. If they do, then this pr should cover historical data too, no?

Hmm. The short answer is no, they don't seem to have it **directly** - but then again they don't have ARSBCH either and in this PR, the code downloads BCHUSD to get from ARSUSD -> BCH (via ARSUSD).

I looked at the API on their website. It's kind of possible but you need to go through USD again first. This endpoint: https://api.yadio.io/hist/365/ARS  <-- gives you last 365 daily average rate of ARSBTC and also USDBTC.  You can get from ARS -> USD that way, and then you would need history data on USDBCH, and can merge the data sets to get ARS -> BCH historically.

So it's.. **possible**.. just messy..

**Comment by emergent-reasons** [comment]
Sounds like historical will be shaky. Better go ahead with this one. I have also tested it locally and it works. Support merging if @cculianu is ok with the quality.

**Comment by fpelliccioni** [comment]
Good job guys, congratulations and thanks!
❤️

**Comment by mhsmith** [comment]
This feature has now been released in the Electron Cash Android app (version 4.2.5-5).

**Pull request merged**